### DISCUSSION TITLE
wxMenus that are NOT hanging off of a menubar

### DISCUSSION THREAD

**Discussion by robertroessler** [wxFormBuilder/wxFormBuilder#848] [Q&A] [answered]
I think this is the right place - I had imprudently started a discussion under Issues previously, but the community was tolerant and helpful even so.
I need to define several menus that will only be used as context menus - but wxFB (both 4.2.1 and a recent nightly) only lets me define one under a menubar (or a dialog).
What's weird is that I actually have one that is at the outmost (frame in my case) level, the way I want - but I have no idea how I created it there, since wxFB always blocks me from creating (or pasting) a menu at this outer level... the only places it allows are under a menubar, or as part of a dialog (why would I do this?).
Since wanting to define menus [only] for use as context popups doesn't seem weird, I assume I am missing something key about how I should be using wxFB?
BTW, neat app overall, and it has definitely helped me in porting an app away from an MFC Windows-only dialog package to a new life as a wxWidgets frame-based app.

**Comment by sodevel** [comment, marked as answer]
Most `wxWindow` based elements allow to attach exactly one `wxMenu` item to them to be used for your desired purpose, as context menu. This also adds a right click event handler that pops up the menu, this cannot be disabled (i never used that feature myself and discovered it only recently).

You cannot add another `wxMenu` to your frame because it has already one attached.

**Comment by robertroessler** [comment]
All right, thanks - my use case still doesn't seem unusual - what if, for instance, you have multiple different context menus for a given window, that may be selected by some additional logic (like seeing that exactly 1 or more than 1 items are currently selected)?
Since _any_ wxMenu can be "popped up", then how about just hanging some of these "extra" menus off of a disabled menubar? Trying...
Well, wxFB is clever enough to block me from doing what I want anywhere _under_ my existing wxFrame... but I had no trouble creating a new menubar at the highest level (project), so I will go ahead and create my gallery of menus under that, and then pop them up wherever I want to (which will hopefully work). ;)

EDIT: well, I did get this to work... apparently, the single "floating" menu I was using as a popup was allowed to exist because it filled the frame's _single_ slot for a menu (sigh)... but I can indeed hang as many menus as I wish for my "gallery" off of the frame's menubar - and not have them in my face, merely by setting their labels to _empty_.

As a detail, since the individual menus on the menubar are not done as derived classes (which I don't care about), then we have a "name scope" issue, where I have to have what are logically the "same" menu item across multiple menus append "2", "3", _etc_ to stop the "redefinition" errors - BUT they are of course still able to use the _same_ ids, so the same handler code is invoked and so on. :)

I hope nobody on the project sees this as a challenge and tries to block this approach. ;)

But really, because of the concept of popped-up context menus, it seems to make sense to loosen up the restrictions on where [stand-alone] menus can be defined, just sayin' - thanks again for responding.

**Comment by baylejs** [comment]
I use wxFormBuilder to build and generate 95% of the UI code, some things you may need to do yourself, like context sensitive pop-up menus, which I use in my app.  I have several pop-up menus associated with various objects defined in wxFormbuilder. In the inherited class init section I add this to the inherited class to define the menu, repeat for each menu you need:

self.shapesCheckListPopupMenu = wx.Menu()
		item = self.shapesCheckListPopupMenu.Append(-1, 'View/Edit Spreadsheet')
		self.Bind(wx.EVT_MENU, self.onShapesCheckListPopupMenuSelected, item)
		item = self.shapesCheckListPopupMenu.AppendSeparator()
		item = self.shapesCheckListPopupMenu.Append(-1, 'Create Property Maps from Selected Shapes')
		self.Bind(wx.EVT_MENU, self.onShapesCheckListPopupMenuSelected, item)
		item = self.shapesCheckListPopupMenu.AppendSeparator()
		item = self.shapesCheckListPopupMenu.Append(-1, 'Delete Selected Items...')
		self.Bind(wx.EVT_MENU, self.onShapesCheckListPopupMenuSelected, item)
		self.m_shapesCheckList.Bind(wx.EVT_CONTEXT_MENU, self.onShapesCheckListRightClick)

Then you define the method to handle the events:

	def onShapesCheckListPopupMenuSelected(self, event):
		item = self.shapesCheckListPopupMenu.FindItemById(event.GetId())
		text = item.GetItemLabelText()
		if text == 'View/Edit Spreadsheet':
			self.onViewTableWindow(event)
		if text == 'Create Property Maps from Selected Shapes':
			self.onConvertShapeToContours(event)
		if text == 'Delete Selected Items...':
			self.onDeleteSelectedItems(event)


Then you define each of those methods:

	def onViewTableWindow( self, event ):
		dataDF = pd.DataFrame()
		checkListSelections = None
		selectionString = ''
		if self.currentCheckList == 'wells':
			checkListSelections = self.m_wellsCheckList.GetSelections()
			selectionString = self.m_wellsCheckList.GetString(checkListSelections[0])
		elif self.currentCheckList == 'maps':
			checkListSelections = self.m_mapsCheckList.GetSelections()
			selectionString = self.m_mapsCheckList.GetString(checkListSelections[0])
		elif self.currentCheckList == 'shapes':
			checkListSelections = self.m_shapesCheckList.GetSelections()
			selectionString = self.m_shapesCheckList.GetString(checkListSelections[0])

Hope this helps, I am by no means an expert.

Jeff

**Comment by robertroessler** [comment]
Yup, thanks, this is basically what I did in the old Windows/MFC C++ app - there, all of the basic "forms" and fixed UI elements - like buttons - were defined using the Visual Studio resource editor (which was reasonably satisfactory, as far as it went).
But then I did what you are doing with all of my context menus - they were all built on the fly using code (sometimes based on current state)... but this was a lot of code that not easy to read or maintain, so I have been pleased with wxFB's being able to let me define pretty much the entire UI - and _it_ then writes the code! ;)
In particular, my current "experiment" is letting wxFB construct things like menus and frames and dialog for me, and then the most I need to do when I use them is possibly enable or disable a few elements.
So I have only been "offering suggestions" ;) in this question, as I believe the current model of "one frame (or whatever), one context menu" is a bit limiting... and as I said in my previous thread entry, I found a "dodge" to get around this with wxFB.
